### PR TITLE
Introduce preflight checks for our testsuite to improve certainty regarding to what will be tested

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

This tries to improve the situation for two scenarios where tests may be skipped unintentionally. The cases are:

1) Side-effects from components in the cluster or from other tests, which change the cluster-state unintentionally in a way which may lead to skips on tests afterwards. For instance a test can mark workers as unschedulable. As a consequece all migration tests would be skipped if there are not enough workers left. So no migration tests would be run, but the test suite would be green.
2) Expectations from the test executor may not be in sync with what the cluster provides. One wants to check if kubevirt runs on the specified dual-stack cluster, but the cluster is not even dual-stack.

Even after this PR the only way to be absolutely sure that everyting got tested which was expected, is looking at the skips in the resulting junit XML. Skip and Focus commands can always lead to wrong results if one is not careful.

To improve the situation the following is done:

1) To detect unwanted side-effects while the tests are running, the cluster initally does a capability discovery of the culster. Skips later on will not only check how the current situation is (e.g. not enough nodes for migration tests), it will also check if the initial discovery had the same opinion. If we had initially enough nodes, but that changed for migration tests, the tests will not be skipped, they will fail.

The auto-discovered values will always be printed before every test run and look like this:

```
{
  "kubernetesVersion": "1.17.6",
  "dualNetworkStack": false,
  "minimumSchedulableNodes": 2,
  "minimumNodesWithCPUManager": 1,
  "isKind": false
}
```

2) It is possible to provide via `-cluster-profile` a file with the same format like the auto-discovery output. The test suite will compare the discovered values with the expectations from the test executor and fail the suite if they differ in an important way.

An example output where the tests immediately abort, due to differing expectations regarding to dual-stack:

```
Discovered cluster:
{
  "kubernetesVersion": "1.17.6",
  "dualNetworkStack": false,
  "minimumSchedulableNodes": 2,
  "minimumNodesWithCPUManager": 1,
  "isKind": false
}
Expected cluster:
{
  "dualNetworkStack": true
}


Failure [0.036 seconds]
[BeforeSuite] BeforeSuite 
tests/tests_suite_test.go:65

  [IsDualStack was answered with false, expected true]
```

Finally the `-skip-dual-stack-tests` flag, introduced in #4139 is still recognized. However it should be replaced by profile-files for each of our kubevirtci clusters.

If there is no `-cluster-profile` provided, the suite will try to run as many tests as it can, based on what it discovered.

Every property in the cluster-profile file is optional. If it is not specified, the test-suite will use the autodiscovered values as base to determine if skips are legitimate.

**Follow Up Actions**

 * Each kubevirtci cluster should provide a cluster-profile file
 * We can add minimal-expectations to our conformance tests (to ensure that e.g. migrations get tested), to better control what full conformance means

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by rmohr** [comment]
The PR is based on #4051 . Only the last 6 PRs are relevant.

File: tests/utils.go [hunk 1]
```diff
@@ -4579,6 +4579,9 @@ func SkipTestIfNoCPUManager() {
 		}
 	}
 	Skip("no node with CPUManager detected")
+	if !HasFeature(virtconfig.CPUManager) {
```
Comment (AlonaKaplan) [inline tests/utils.go]: Shouldn't it be moved to the beginning of the method? In this stage the test is already skipped.

File: tests/checks/preflight.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			errors.errors = append(errors.errors, fmt.Errorf("OpenShiftMajorVersion was answered with %v, expected %v", *DiscoveredClusterProfile.OpenShiftMajorVersion, *expected.OpenShiftMajorVersion))
+		}
+	}
+	if expected.KubernetesVersion != nil {
+		if !reflect.DeepEqual(*expected.KubernetesVersion, *DiscoveredClusterProfile.KubernetesVersion) {
+			errors.errors = append(errors.errors, fmt.Errorf("KubernetesVersion was answered with %v, expected %v", *DiscoveredClusterProfile.KubernetesVersion, *expected.KubernetesVersion))
+		}
+	}
+	if expected.DualNetworkStack != nil {
+		if !reflect.DeepEqual(*expected.DualNetworkStack, *DiscoveredClusterProfile.DualNetworkStack) {
```
Comment (AlonaKaplan) [inline tests/checks/preflight.go:120]: Why `DeepEqual` is needed to compare booleans? Same for the other compares of simple types.

File: tests/checks/preflight.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+type DiscoveryError struct {
+	errors []error
+}
+
+func (d *DiscoveryError) Error() string {
+	return fmt.Sprintf("%v", d.errors)
+}
+
+func VerifyClusterExpectations() error {
+	if len(flags.ClusterProfilePath) == 0 {
```
Comment (AlonaKaplan) [inline tests/checks/preflight.go:90]: I case the cluster doesn't have a profile, don't we want to compare the current `DiscoveredClusterProfile` to the first discovered one?

File: tests/checks/preflight.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		log.DefaultLogger().Reason(err).Critical("Could not find the provided cluster profile file.")
+	}
+	if err := yaml.NewYAMLOrJSONDecoder(reader, 1024).Decode(expected); err != nil {
+		log.DefaultLogger().Reason(err).Critical("Could not decode the provided cluster profile file.")
+	}
+
+	errors := &DiscoveryError{}
+	if expected.OpenShiftMajorVersion != nil {
+		if !reflect.DeepEqual(*expected.OpenShiftMajorVersion, *DiscoveredClusterProfile.OpenShiftMajorVersion) {
```
Comment (AlonaKaplan) [inline tests/checks/preflight.go:110]: The method assumes `DiscoveredClusterProfile` was already initialized. I would change `IntrospectCluster` to be a private method and invoke at the beginning of `IntrospectCluster`,

File: tests/utils.go (hunk 1 above)
Comment (rmohr) [reply tests/utils.go]: Definitely. :+1:

File: tests/checks/preflight.go (hunk 2 above)
Comment (rmohr) [reply tests/checks/preflight.go:120]: Just copy-pase. Will change it.

File: tests/checks/preflight.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+			errors.errors = append(errors.errors, fmt.Errorf("OpenShiftMajorVersion was answered with %v, expected %v", *DiscoveredClusterProfile.OpenShiftMajorVersion, *expected.OpenShiftMajorVersion))
+		}
+	}
+	if expected.KubernetesVersion != nil {
+		if !reflect.DeepEqual(*expected.KubernetesVersion, *DiscoveredClusterProfile.KubernetesVersion) {
+			errors.errors = append(errors.errors, fmt.Errorf("KubernetesVersion was answered with %v, expected %v", *DiscoveredClusterProfile.KubernetesVersion, *expected.KubernetesVersion))
+		}
+	}
+	if expected.DualNetworkStack != nil {
```
Comment (AlonaKaplan) [inline tests/checks/preflight.go:119]: dual_stack_cluster.go can be removed now.

**Review by oshoval — COMMENTED** [review]
Nice,
can you catch the offender test ? (unless you already doing it)
for example if each failure of a test it would check saved status versus the actual status,
so if there are rare bugs that change the cluster status we would find them easier

File: tests/checks/preflight.go [hunk 6]
```diff
@@ -38,39 +39,75 @@ type ClusterProfile struct {
...
+	discovered.IsKind = &isKind
+	return discovered
+}
 
+func (c *clusterIntrospector) loadExpectedClusterProfile() ClusterProfile {
+	expected := &ClusterProfile{}
+	if len(flags.ClusterProfilePath) > 0 {
+		reader, err := os.Open(flags.ClusterProfilePath)
+		if err != nil {
+			log.DefaultLogger().Reason(err).Critical("Could not find the provided cluster profile file.")
```
Comment (enp0s3) [inline tests/checks/preflight.go:110]: Shouldn't we be more strict here and exit the suite with error that we couldn't open the provided cluster profile?

File: tests/checks/preflight.go [hunk 7]
```diff
@@ -81,15 +118,7 @@ func (d *DiscoveryError) Error() string {
...
-
-	// XXX this flags needs to go
-	// it makes dual stack the default assumption if no profile file is provided
-	if len(flags.ClusterProfilePath) == 0 {
-		deprecatedIsDualStack := !flags.SkipDualStackTests
-		expected.DualNetworkStack = &deprecatedIsDualStack
-	}
+func (c *clusterIntrospector) verifyClusterExpectations(expected ClusterProfile, DiscoveredClusterProfile ClusterProfile) error {
 
 	if len(flags.ClusterProfilePath) > 0 {
```
Comment (enp0s3) [inline tests/checks/preflight.go:129]: Shouldn't it be removed since you already open the file in the `loadExpectedClusterProfile` member function?

**Review by enp0s3 — COMMENTED** [review]
Great work! I have a few questions, please see below

**Pull request closed without merging**